static/js/datasource.js

```diff
@@ -0,0 +1,208 @@
+import * as HTTP from './http.js'
+import EventEmitter from './eventemitter.js'
+
+class DataSource {
+  static DEFAULT_STATE = {
+    page: 1,
+    page_size: 100,
+    sort_key: '',
+    sort_reverse: false,
+    name: ''
+  }
+  constructor(opts) {
+    this._opts = Object.assign(
+      {
+        autoReloadTimeout: 5000,
+        useQueryState: true
+      }, opts)
+    this._reloadTimer = null
+    this._events = new EventEmitter()
+    this._items = []
+    this._filteredCount = 0
+    this._itemCount = 0
+    this._pageCount = 0
+    this._totalCount = 0
+    this._setState()
+    let cachedState
+    this._lastLoadedUrl = ''
+    if (this._opts.useQueryState) {
+      if (cachedState = window.sessionStorage.getItem(this._cacheKey)) {
+        try {
+          cachedState = JSON.parse(cachedState)
+          this._setState(cachedState)
+        } catch(e) {
+          console.error(`Failed to load cached query state: ${e}`)
+        }
+      }
+      this._setStateFromQuery()
+      window.addEventListener('hashchange', evt => this._setStateFromQuery())
+    }
+  }
+
+  _setStateFromQuery() {
```

Comment:
any reason you're not using the private prefix `#`? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

Reply:
I did at first, then I saw something related that wasn't fully supported so I changed. Dunno what now.